### DISCUSSION TITLE
Tags handling rewrite

### DISCUSSION THREAD

**Discussion by nicosomb** [wallabag/wallabag#6784] [Ideas]
Today, our tags system is really not perfect. 

## Today

Here is the database model: 
#### wallabag_tag

The table used to store tags. 

- id
- label
- slug
#### wallabag_entry 

The table used to store entries. 

- id
- url
- user_id
- and so many other fields
#### wallabag_entry_tag

The table used for the relation between a tag and an entry. 

- entry_id
- tag_id 

### Main problem

**The tags aren’t user-related.** 

It can cause performance issues.  
One example on app.wallabag.it for the tag `politics`: there are 46.122 entries with this tag. So, each time a user want to add / remove this tag from an entry, we embed too much entries. There is an issue for this bug: https://github.com/wallabag/wallabag/issues/4998

#### wallabag_tag

- id
- **user_id**
- label
- slug
#### wallabag_entry 

- id
- url
- user_id

#### wallabag_entry_tag

- entry_id
- tag_id 

#### Migrations 

We have to update `wallabag_tag` to add a relation to the `wallabag_user` table. 

Then, we have to update the existing content.  
For each tag, we have to update the new `user_id` field with the `user_id` of the related entry. 

## Bugs related to tags

Here are the issues which won’t be fixed until we update the database schema (feel free to link other issues related to tags in comments below): 

- https://github.com/wallabag/wallabag/issues/6699
- https://github.com/wallabag/wallabag/issues/6534
- https://github.com/wallabag/wallabag/issues/6532
- https://github.com/wallabag/wallabag/issues/4709
- https://github.com/wallabag/wallabag/issues/4998 
- https://github.com/wallabag/wallabag/issues/4239
- https://github.com/wallabag/wallabag/issues/719

We also have features requests related to tag (autocompletion for example). 
I think that these features should be implemented only when we’ll do the tags handling rewrite. 

## Feel free to participate to this rewrite 

The “Tomorrow” section is only my first idea to improve the tags handling. Feel free to comment if you have other ideas 👇

**Comment by Kdecherf** [comment]
The rewrite would also help third-party applications like the Android one as there's no easy way to clean up removed/unused tags without a full resync.

**Comment by Kdecherf** [comment]
Here are some thoughts I had recently on tags:

I agree with the need to update the database model to actually attach tag entities to users, it would simplify housekeeping and code a lot imo. It should be noted that this may not completely fix the performance issue that can be seen on wallabag.it, there may be another refacto needed here.

Having tags associated to users would let them do several new things:
- Create new tags without any entry
- Customize label (_I'm doing this to show emojis in the label but keep a specific slug_)
- Add colors to tags (_this is a thing I heavily rely on, see [this blog post](https://kdecherf.com/blog/2021/05/15/wallabag-add-colors-to-tags-with-pure-css/)_)

**Comment by j0k3r** [comment]
I'm all in for that too.
Tags must not be global because people can have whatever name they want to a tag and it shouldn't affect someone else.

The new schema looks ok to me.
One thing to be careful is the slug. It must now include the username inside it (or the user id) to ensure uniqueness accross the table. Mostly because we are using the slug to access the list of entries for a tag, like `https://app.wallabag.it/tag/list/medium`.

Or maybe we need a different solution about URL, maybe using the database id instead?

**Comment by yguedidi** [comment]
Created this #8909
